linux/GPU.c

```diff
@@ -0,0 +1,256 @@
+            continue;
+
+         if (String_startsWith(buf, "drm-client-id:")) {
+            char *endptr;
+
+            if (sstate == SECST_NEW) {
+               struct client_id* new;
+
+               assert(client_id != INVALID_CLIENT_ID);
+
+               new = xMalloc(sizeof(*new));
+               *new = (struct client_id) {
+                  .id = client_id,
+                  .pdev = pdev,
+                  .next = parsed_ids,
+               };
+               pdev = NULL;
+
+               parsed_ids = new;
+            }
+
+            sstate = SECST_UNKNOWN;
+
+            errno = 0;
+            client_id = strtoull(buf + strlen("drm-client-id:"), &endptr, 10);
+            if (errno || *endptr != '\n')
+               client_id = INVALID_CLIENT_ID;
+         } else if (String_startsWith(buf, "drm-pdev:")) {
+            const char* p;
+            char* q;
+
+            p = buf + strlen("drm-pdev:");
+
+            while (isspace((unsigned char)*p))
+               p++;
+
+            q = strchr(p, '\n');
+            if (q)
+               *q = '\0';
+
+            assert(!pdev || String_eq(pdev, p));
+            if (!pdev)
+               pdev = xStrdup(p);
+         } else if (String_startsWith(buf, "drm-engine-")) {
+            const char* delim;
+            const char* engineStart;
+            char* endptr;
+            unsigned long long int value;
+
+            if (sstate == SECST_DUPLICATE)
+               continue;
+
+            if (String_startsWith(buf, "drm-engine-capacity-"))
+               continue;
+
+            delim = strchr(buf, ':');
+            engineStart = buf + strlen("drm-engine-");
+
+            errno = 0;
+            value = strtoull(delim + 1, &endptr, 10);
+            if (errno == 0 && String_startsWith(endptr, " ns")) {
+               if (sstate == SECST_UNKNOWN) {
+                  if (client_id != INVALID_CLIENT_ID && !is_duplicate_client(parsed_ids, client_id, pdev))
+                     sstate = SECST_NEW;
+                  else
+                     sstate = SECST_DUPLICATE;
+               }
+
+               if (sstate == SECST_NEW) {
+                  new_gpu_time += value;
+                  update_machine_gpu(lpt, value, engineStart, delim - engineStart);
+               }
+            }
+         }
+      } /* finished parsing lines */
+
+      fclose(fp);
+
+      if (sstate == SECST_NEW) {
+         struct client_id* new;
+
+         assert(client_id != INVALID_CLIENT_ID);
+
+         new = xMalloc(sizeof(*new));
+         *new = (struct client_id) {
+            .id = client_id,
+            .pdev = pdev,
+            .next = parsed_ids,
+         };
+         pdev = NULL;
+
+         parsed_ids = new;
+      }
+
+      free(pdev);
+   } /* finished parsing fdinfo entries */
+
+   if (new_gpu_time > 0) {
+      const Machine* host = lp->super.super.host;
+      unsigned long long int gputimeDelta;
+      uint64_t monotonictimeDelta;
```

Comment:
A small nitpick on the naming: `monotonictimeDelta` or `monotonicTimeDelta`, which is more readable?